### ISSUE TITLE
Programming Model : Basics

### ISSUE THREAD

**Issue by asavaritayal** [Azure/azure-functions-python-worker#2]
## Script File 
Python support for Functions will follow the existing scripting model and support a run file of **.py format**. We'll use the workflow below to determine which file to run:

_If_ the **"scriptFile"** attribute is set in **function.json**
_Else if_ the project contains a single **.py file**
_Else if_ **main.py** is present
_Else_ throw an error
## Entry Point 
Instead of wrapping everything in a class, we'll define a  **main()** method for the runtime to invoke during an execution.  We'll use the workflow below to determine which function to run:

_If_ the **"entryPoint"** attribute is set in **function.json**
_Else if_ **main()** is present
_Else_ throw an error

## Function format
Any references to external modules  will be included at the beginning of the file using the **import** keyword. Additional classes and/or helper functions can also be added to the same file.

```python
# main.py
import os
def main():
    pass
```

Since only a single Python process can exist per function app, it is recommended to implement main()  as an asynchronous coroutine using the **async def** statement.
```python
# Would be run with asyncio directly
async def main():
    await some_nonblocking_socket_io_op()
```
If the `main()` function is synchronous (no `async` qualifier) we automatically run it in an asyncio thread-pool:
```python
# Would be run in an asyncio thread-pool
def main():
    some_blocking_socket_io()
```
This approach gives the user a choice of what libraries they want to use and async/await becomes an *opt-in*.

**Comment by 1st1** [comment]
> # Entry Point
>
>Instead of wrapping everything in a class, we'll define a main() method for the runtime to invoke during an execution. We'll use the workflow below to determine which function to run:
>
>If the "entryPoint" attribute is set in function.json
>Else if the script contains a single function
>Else if main() is present
>Else throw an error

"Else if the script contains a single function" —this can be problematic. It's hard to detect if a module defines one function, or if it imports it.  We would need to parse the file and analyze the AST to implement this.

Instead, I suggest to shorten this to:

* If the "entryPoint" attribute is set in function.json
* Else if `main()` is present
* Else throw an error

**Comment by elprans** [comment]
What about also looking for `main()` in `__init__.py`, if present?  Or maybe make that the default `scriptFile` altogether.  Seems a bit more pythonic to me.

**Comment by 1st1** [comment]
> What about also looking for main() in __init__.py, if present? Or maybe make that the default scriptFile altogether. Seems a bit more pythonic to me.

+1.

@brettcannon would do you think about this? cc @asavaritayal @christopheranderson @pragnagopa

**Comment by brettcannon** [comment]
SGTM

**Comment by 1st1** [comment]
Elvis, can you take a look at it then?

**Comment by elprans** [comment]
It seems like the worker doesn't actually have a say in selecting the default `scriptFile`.  Instead, the host picks the file with the appropriate extension, if it's the only file in the directory; or the file called `run`, or `index.js`.

https://github.com/Azure/azure-functions-host/blob/1f243e9febc4d431af3f0341bc8af74975d51659/src/WebJobs.Script/Host/ScriptHost.cs#L1258-L1280

@christopheranderson: we _could_ add `__init__.py` to the list of "special" filenames, but that doesn't seem scalable to me.  I think we should push the logic down to the worker, where the language specifics are easier to handle.

**Comment by asavaritayal** [comment]
From the offline discussion, we'll adopt `__init__.py` as the default script file for Python.

Let's update the docs - [https://pythondeveloperguide.azurewebsites.net/](https://pythondeveloperguide.azurewebsites.net/)before closing the issue.
